### PR TITLE
test(perf): gradual throughput grow test

### PR DESCRIPTION
**Pull request by juliayakovlev**
### Testing

- [ ]

### PR pre-checks (self review)

- [ ] I added the relevant `backport` labels
- [ ] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

### CODE REVIEW

File: performance_regression_gradual_grow_throughput.py [hunk 1]
```diff
@@ -14,29 +14,34 @@ class CSPopulateDistribution(Enum):
 
 class PerformanceRegressionGradualGrowThroughutTest(PerformanceRegressionTest):  # pylint: disable=too-many-instance-attributes
 
+    max_ops = 238_000
+    start_ops = int(max_ops * 0.25)  # 25% from max_ops
+    throttle_step = int(max_ops * 0.5)
```
Comment (roydahan) [inline performance_regression_gradual_grow_throughput.py:19]: Let's take these out to the test yaml

Comment (roydahan) [reply performance_regression_gradual_grow_throughput.py:19]: The throttle steps should be:
0.25, 0.5, 0.75, 0.95, max (no throttle).

File: test-cases/performance/perf-regression-gradual-throughput-grow.yaml [hunk 2]
```diff
@@ -1,34 +1,62 @@
...
+  "cassandra-stress read no-warmup cl=QUORUM duration=5m -mode cql3 native -rate 'threads=$threads throttle=$throttle/s' -col 'size=FIXED(1024) n=FIXED(1)' -pop 'dist=gauss(108333335..216666668,162500002,542209)'",
+  "cassandra-stress read no-warmup cl=QUORUM duration=5m -mode cql3 native -rate 'threads=$threads throttle=$throttle/s' -col 'size=FIXED(1024) n=FIXED(1)' -pop 'dist=gauss(216666669..325000002,270833336,542209)'",
+#  "cassandra-stress read no-warmup cl=QUORUM duration=15m -mode cql3 native -rate 'threads=$threads throttle=$throttle/s' -col 'size=FIXED(1024) n=FIXED(1)' -pop 'dist=gauss(325000003..433333336,379166670,542209)'",
+#  "cassandra-stress read no-warmup cl=QUORUM duration=15m -mode cql3 native -rate 'threads=$threads throttle=$throttle/s' -col 'size=FIXED(1024) n=FIXED(1)' -pop 'dist=gauss(433333337..541666670,487500004,542209)'",
+#  "cassandra-stress read no-warmup cl=QUORUM duration=15m -mode cql3 native -rate 'threads=$threads throttle=$throttle/s' -col 'size=FIXED(1024) n=FIXED(1)' -pop 'dist=gauss(541666671..650000004,595833338,542209)'"
```
Comment (roydahan) [inline test-cases/performance/perf-regression-gradual-throughput-grow.yaml:23]: Let's use 20m per step.

File: performance_regression_gradual_grow_throughput.py (hunk 1 above)
Comment (roydahan) [reply performance_regression_gradual_grow_throughput.py:19]: All with 1K threads.

File: performance_regression_gradual_grow_throughput.py [hunk 3]
```diff
@@ -182,45 +204,55 @@ def run_gradual_increase_load(self, stress_cmd_templ,
...
+    def warmup_cache(self, stress_cmd_templ):
+        stress_queue = []
+        for stress_cmd in stress_cmd_templ:
+            params = {"round_robin": True, "stats_aggregate_cmds": False}
+            stress_cmd_to_run = stress_cmd.replace("$threads", "1000")
+            # Run c-s command without throttle
+            stress_cmd_to_run = stress_cmd_to_run.replace("$throttle", "")
+            # TODO: get "n" from c-s command or do it configurable
+            stress_cmd_to_run = stress_cmd_to_run.replace(
+                "duration=20m", "n=216666669" if "gauss" in stress_cmd_to_run else "n=20000000")
```
Comment (roydahan) [inline performance_regression_gradual_grow_throughput.py:249]: I'm not sure I understand the else part: "else "n=20000000"".
What is this 20M?

File: performance_regression_gradual_grow_throughput.py [hunk 4]
```diff
@@ -182,45 +204,55 @@ def run_gradual_increase_load(self, stress_cmd_templ,
...
+            # For mixed workflow run read warm up
```
Comment (roydahan) [inline performance_regression_gradual_grow_throughput.py:250]: One more thing is missing, to replace the QUORUM with ALL.

File: configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# Define load ops for steps
+perf_max_ops: {"read": 794000, "write": 405000, "mixed": 530000}
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml:2]: @roydahan 
Yes, I will remove it later. There are different modules that use this. So I did not touch it for now. Just want to see how we will advance with new aproach

File: configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+perf_threads: {"read": 1000, "write": 700, "mixed": 900}
+perf_throttle_steps: [150000, 300000, 450000, 600000, 700000, "MAX"]  # percents from max ops: 25%, 50%, 75%, 90%, max_ops
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml:4]: @roydahan do you mean 
```
perf_throttle_steps: [750000, "MAX"]
```
?

File: performance_regression_gradual_grow_throughput.py (hunk 3 above)
Comment (juliayakovlev) [reply performance_regression_gradual_grow_throughput.py:249]: It is for read test. We run smaller data set for read

File: performance_regression_gradual_grow_throughput.py (hunk 4 above)
Comment (juliayakovlev) [reply performance_regression_gradual_grow_throughput.py:250]: right, missed that

Comment (juliayakovlev) [reply performance_regression_gradual_grow_throughput.py:250]: fixed

File: configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (roydahan) [reply configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml]: Yes

File: configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (roydahan) [reply configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml:2]: ok.

File: performance_regression_gradual_grow_throughput.py (hunk 3 above)
Comment (roydahan) [reply performance_regression_gradual_grow_throughput.py:249]: but it's not 20M, it's 5M for each command, isn't it?

File: configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml]: done

**Review by roydahan — CHANGES_REQUESTED** [review]

File: configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (roydahan) [reply configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml]: I meant:
perf_throttle_steps: [150000, 300000, 450000, 600000, 750000, "MAX"]

File: jenkins-pipelines/performance_staging/perf-regression-scylla-gradual-throughput-grow-uniform.jenkinsfile
Comment (fruch) [inline jenkins-pipelines/performance_staging/perf-regression-scylla-gradual-throughput-grow-uniform.jenkinsfile:1]: I think we can remove the gauss/uniform from the name of the pipeline, as we gonna pick only one of them ?

Comment (roydahan) [reply jenkins-pipelines/performance_staging/perf-regression-scylla-gradual-throughput-grow-uniform.jenkinsfile:1]: in fact we didn't choose any of them. we're using seq.

File: configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# Define load ops for steps
+perf_max_ops: {"read": 794000, "write": 405000, "mixed": 530000}
+perf_threads: {"read": 620, "write": 700, "mixed": 1900}
```
Comment (roydahan) [inline configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml:3]: for write I asked to use "400".
```suggestion
perf_threads: {"read": 620, "write": 400, "mixed": 1900}
```

File: configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml [hunk 8]
```diff
@@ -0,0 +1,4 @@
...
+perf_gradual_throttle_steps: {"read": '150000', '300000', '450000', '600000', '700000', 'unthrotteled', "mixed": ['50000', '150000', '300000', '450000', 'unthrotteled']}  # where every value is in ops
```
Comment (roydahan) [inline configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml:4]: "write": '200000', '300000', '400000', 'unthrotteled'

Not sure why the format is different between read and mixed.

File: test-cases/performance/perf-regression-gradual-throughput-grow.yaml [hunk 9]
```diff
@@ -1,34 +1,64 @@
...
 
-instance_type_loader: 'c5n.9xlarge'
-instance_type_monitor: 't3.large'
-instance_type_db: 'i3.4xlarge'
 
-user_prefix: 'perf-1tb-scylla-gradual-grow-throughput'
-space_node_threshold: 644245094
-ami_id_db_scylla_desc: 'VERSION_DESC'
+instance_type_db: 'i4i.4xlarge'
+instance_type_loader: 'c7i.12xlarge'
```
Comment (roydahan) [inline test-cases/performance/perf-regression-gradual-throughput-grow.yaml:41]:
```suggestion
instance_type_loader: 'c7i.8xlarge'
```

Didn't test, but should be ok.

File: jenkins-pipelines/performance_staging/perf-regression-scylla-gradual-throughput-grow-gauss.jenkinsfile [hunk 10]
```diff
@@ -6,9 +6,9 @@ def lib = library identifier: 'sct@snapshot', retriever: legacySCM(scm)
 perfRegressionParallelPipeline(
     backend: "aws",
     aws_region: "us-east-1",
-    test_name: "performance_regression_gradual_grow_throughput.PerformanceRegressionGradualGrowThroughutTest",
-    test_config: '''["test-cases/performance/perf-regression-gradual-throughput-grow.yaml", "configurations/cs-gauss-distribution.yaml"]''',
-    sub_tests: ["test_write_gradual_increase_load","test_read_gradual_increase_load", "test_mixed_gradual_increase_load"],
+    test_name: "performance_regression_gradual_grow_throughput.PerformanceRegressionGradualGrowThroughputTest",
+    test_config: '''["test-cases/performance/perf-regression-gradual-throughput-grow.yaml", "configurations/cs-gauss-distribution.yaml", "configurations/performance/cassandra_stress_gradual_load_steps_enterprise.yaml", "configurations/disable_kms.yaml", "configurations/tablets_disabled.yaml", "configurations/disable_speculative_retry.yaml"]''',
```
Comment (roydahan) [inline jenkins-pipelines/performance_staging/perf-regression-scylla-gradual-throughput-grow-gauss.jenkinsfile:10]: You should remove the "configurations/cs-gauss-distribution.yaml" from the config.
I don't think we use it anywhere (or should use).

File: performance_regression_gradual_grow_throughput.py [hunk 11]
```diff
@@ -128,34 +108,98 @@ def preload_data(self, compaction_strategy=None):
...
+        throttle_steps = self.params["perf_gradual_throttle_steps"]
+        if sub_type not in throttle_steps:
+            TestFrameworkEvent(source=self.__class__.__name__,
+                               message=f"Throttle steps for '{sub_type}' test is not defined in 'perf_gradual_throttle_steps' parameter",
+                               severity=Severity.CRITICAL).publish()
+        workload = getattr(PerformanceTestWorkload, sub_type.upper())
+
+        for throttle_step in throttle_steps[sub_type]:
+            if sub_type == "write":
+                self.drop_keyspace()
```
Comment (roydahan) [inline performance_regression_gradual_grow_throughput.py:169]: The "drop" should be in the end of the step, not in the begining, just before the wait for compactions, so we won't wait for compactions.

**Comment by juliayakovlev** [comment]
Close with https://github.com/scylladb/scylla-cluster-tests/pull/8734

**Pull request closed without merging**